File: src/main/java/net/sf/rails/game/model/SpecialPropertiesModel.java
package net.sf.rails.game.model;

import net.sf.rails.common.LocalText;
import net.sf.rails.common.ReportBuffer;
import net.sf.rails.game.Bonus;
import net.sf.rails.game.PublicCompany;
import net.sf.rails.game.RailsOwner;
import net.sf.rails.game.financial.Bank;
import net.sf.rails.game.special.LocatedBonus;
import net.sf.rails.game.special.SpecialProperty;
import net.sf.rails.game.state.Change;
import net.sf.rails.game.state.Observable;
import net.sf.rails.game.state.SetChange;
import net.sf.rails.game.state.PortfolioSet;
import net.sf.rails.game.state.Triggerable;

public class SpecialPropertiesModel extends RailsModel implements Triggerable {

    public final static String ID = "SpecialPropertiesModel";

    private final PortfolioSet<SpecialProperty> specialProperties;

    private SpecialPropertiesModel(RailsOwner parent) {
        super(parent, ID);
        // specialProperties have the Owner as parent directly
        specialProperties = PortfolioSet.create(parent, "specialProperties", SpecialProperty.class);
        // so make this model updating
        specialProperties.addModel(this);
        // and add it as triggerable
        specialProperties.addTrigger(this);
    }
    
    public static SpecialPropertiesModel create(RailsOwner parent) {
        return new SpecialPropertiesModel(parent);
    }
    
    @Override
    public RailsOwner getParent() {
        return (RailsOwner)super.getParent();
    }

    PortfolioSet<SpecialProperty> getPortfolio() {
        return specialProperties;
    }
    
    // triggerable interface
    
    public void triggered(Observable observable, Change change) {
        
        // checks if the specialproperty moved into the portfolio carries a LocatedBonus

        if (!(change instanceof SetChange)) {
            return;
        }
        
        @SuppressWarnings("rawtypes")
        SetChange sChange = (SetChange)change;
        if (!sChange.isAddToSet()) return;

        SpecialProperty property = (SpecialProperty)sChange.getElement();
        if (getParent() instanceof PublicCompany && property instanceof LocatedBonus) {
            PublicCompany company = (PublicCompany)getParent();
            LocatedBonus locBonus = (LocatedBonus)property;
            Bonus bonus = new Bonus(company, locBonus.getId(), locBonus.getValue(),
                    locBonus.getLocations());
            company.addBonus(bonus);
            ReportBuffer.add(this, LocalText.getText("AcquiresBonus",
                    getParent().getId(),
                    locBonus.getName(),
                    Bank.format(company, locBonus.getValue()),
                    locBonus.getLocationNameString()));
        }
    }

}

File: dotnet/RailsLib.Net/Net/Game/Model/SpecialPropertiesModel.cs
using GameLib.Net.Common;
using GameLib.Net.Game.Financial;
using GameLib.Net.Game.Special;
using GameLib.Net.Game.State;
using System;
using System.Collections.Generic;
using System.Text;

namespace GameLib.Net.Game.Model
{
    public class SpecialPropertiesModel : RailsModel, ITriggerable
    {
        public const string ID = "SpecialPropertiesModel";

        private PortfolioSet<SpecialProperty> specialProperties;

        private SpecialPropertiesModel(IRailsOwner parent) : base(parent, ID)
        {
            // specialProperties have the Owner as parent directly
            specialProperties = PortfolioSet<SpecialProperty>.Create(parent, "specialProperties");
        // so make this model updating
        specialProperties.AddModel(this);
        // and add it as triggerable
        specialProperties.AddTrigger(this);
        }

        public static SpecialPropertiesModel Create(IRailsOwner parent)
        {
            return new SpecialPropertiesModel(parent);
        }

        new public IRailsOwner Parent
        {
            get
            {
                return (IRailsOwner)base.Parent;
            }
        }

        public PortfolioSet<SpecialProperty> Portfolio
        {
            get
            {
                return specialProperties;
            }
        }

        // triggerable interface

        public void Triggered(Observable observable, Change change)
        {

            // checks if the specialproperty moved into the portfolio carries a LocatedBonus

            if (!(change is SetChange<SpecialProperty>)) 
                {
                return;
            }

            //@SuppressWarnings("rawtypes")
            SetChange<SpecialProperty> sChange = change as SetChange<SpecialProperty>;
            if (!sChange.IsAddToSet) return;

            SpecialProperty property = sChange.Element;
            if ((Parent is PublicCompany) && (property is LocatedBonus)) 
                {
                PublicCompany company = (PublicCompany)Parent;
                LocatedBonus locBonus = (LocatedBonus)property;
                Bonus bonus = new Bonus(company, locBonus.Id, locBonus.Value, locBonus.Locations);
                company.AddBonus(bonus);
                ReportBuffer.Add(this, LocalText.GetText("AcquiresBonus",
                        Parent.Id,
                        locBonus.Name,
                        Bank.Format(company, locBonus.Value),
                        locBonus.LocationNameString));
            }
        }
    }
}
